spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/redis/TraceRedisAutoConfiguration.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2013-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.sleuth.instrument.redis;
+
+import brave.Tracing;
+import io.lettuce.core.resource.ClientResources;
+import io.lettuce.core.tracing.BraveTracing;
+
+import org.springframework.beans.BeansException;
+import org.springframework.beans.factory.config.BeanPostProcessor;
+import org.springframework.boot.autoconfigure.AutoConfigureAfter;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.cloud.sleuth.autoconfig.TraceAutoConfiguration;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+/**
+ * {@link org.springframework.boot.autoconfigure.EnableAutoConfiguration
+ * Auto-configuration} enables Redis span information propagation.
+ *
+ * @author Chao Chang
+ * @since 2.2.0
+ */
+@Configuration
+@OnRedisEnabled
+@ConditionalOnBean(Tracing.class)
+@AutoConfigureAfter({ TraceAutoConfiguration.class })
+public class TraceRedisAutoConfiguration {
+
+	@Configuration
+	@ConditionalOnClass(ClientResources.class)
```

Comment:
I'm a bit torn on the conditionality. `TraceLettuceClientResourcesBeanPostProcessor` conditionally configures `ClientResources`  only if no `tracing` is set. If `ClientResources.tracing` is set, `TraceLettuceClientResourcesBeanPostProcessor` just skips configuration without any external visibility that configuration was skipped.

I would propose to add another `@Condition` that introspects a `ClientResources` bean and only then the `TraceLettuceClientResourcesBeanPostProcessor` gets registered. Configuration should happen if `ClientResources.tracing().isEnabled()` returns `false`.

Reply:
Do you mean the code like the following? 
I tried but `TraceLettuceClientResourcesBeanPostProcessor.postProcessAfterInitialization()` was never called for the ClientResources bean created in my test case.
I am not sure what I was missing.

```
public class TraceRedisAutoConfiguration {

	@Configuration
	@ConditionalOnBean(ClientResources.class)
	static class LettuceConfig {

		@Bean
		@Conditional(TracingIsDisabledInLettuceClientResourcesCondition.class)
		static TraceLettuceClientResourcesBeanPostProcessor traceLettuceClientResourcesBeanPostProcessor(
				Tracing tracing) {
			return new TraceLettuceClientResourcesBeanPostProcessor(tracing);
		}

	}

}


class TracingIsDisabledInLettuceClientResourcesCondition extends SpringBootCondition
		implements ConfigurationCondition {

	@Override
	public ConfigurationPhase getConfigurationPhase() {
		return ConfigurationPhase.REGISTER_BEAN;
	}

	@Override
	public ConditionOutcome getMatchOutcome(ConditionContext context,
			AnnotatedTypeMetadata metadata) {
		Assert.isInstanceOf(DefaultListableBeanFactory.class, context.getBeanFactory());
		DefaultListableBeanFactory listableBeanFactory = (DefaultListableBeanFactory) context
				.getBeanFactory();
		ClientResources clientResources = listableBeanFactory
				.getBean(ClientResources.class);

		if (clientResources.tracing().isEnabled()) {
			return ConditionOutcome.noMatch(
					"Tracing in the created ClientResources bean is already enabled.");
		}
		return ConditionOutcome.match();

	}

}

```

Reply:
That looks in the direction how it is supposed to work. Enable debug mode as it should print why the bean was/was not configured.

Reply:
The problem with the above code is that, when we get the `ClientResources ` bean to introspect in TracingIsDisabledInLettuceClientResourcesCondition , the bean will be created and BeanPostProcessors that created before that will be called, at that time,  `TraceLettuceClientResourcesBeanPostProcessor ` is not created yet.
After the `TraceLettuceClientResourcesBeanPostProcessor `  bean is created later, because the creation process of `ClientResources` bean has already finished,  no further `TraceLettuceClientResourcesBeanPostProcessor ` will be called against it.
Would you mind give me some hint on how the following logic can be implemented?
>add another @Condition that introspects a ClientResources bean and only then the TraceLettuceClientResourcesBeanPostProcessor gets registered.

Reply:
I see. I suppose in that case we just need to accept the fact that we cannot use a `@Condition` to introspect `ClientResources`.  I'd suggest then adding logging on `DEBUG` level to `TraceLettuceClientResourcesBeanPostProcessor` for the two cases of configuration and non-configuration. By enabling debug logging, users can at least find some indication on the log, why tracing was configured or why it was not configured by `TraceLettuceClientResourcesBeanPostProcessor`.

Reply:
Debug logs were added.